r/R/dplyr-functions.R

```diff
@@ -280,6 +284,81 @@ nse_funcs$str_trim <- function(string, side = c("both", "left", "right")) {
   Expression$create(trim_fun, string)
 }
 
+nse_funcs$substr <- function(string, start, stop) {
+  assert_that(
+    length(start) == 1,
+    msg = "`start` must be length 1 - other lengths are not supported in Arrow"
+  )
+  assert_that(
+    length(stop) == 1,
+    msg = "`stop` must be length 1 - other lengths are not supported in Arrow"
+  )
+
+  if (start <= 0) {
+    start <- 1
+  }
+
+  if (stop < start) {
+    stop <- 0
+  }
+
+  Expression$create(
+    "utf8_slice_codeunits",
+    string,
+    options = list(start = start - 1L, stop = stop)
```

Comment:
Why do we only subtract 1 from start but not stop?

Reply:
we don't need to subtract 1 from `end` as R counts inclusively (i.e. returned string includes the item at position `end`) whereas C++ counts exclusively (i.e. returned string includes up to the item at position `end` which effectively cancels out the difference due to indexing

Reply:
This deserves an explanatory comment